# Fine-tune Llama 2 for chat & dialogue summarization

Welcome!

Llama 2 is a collection of pretrained and fine-tuned generative text models ranging in scale from 7 billion to 70 billion parameters.

In this Jupyter Notebook, we will learn how to fine-tune the meta-llama/Llama-2-7b-hf, the smallest of the collection, with 7B parameters, using AI Stduio Deep Learning workspace with GPU and MLFlow for tracking the metrics!

## Setup Development Environment

Our first step is to install some extra required packages.

##### Installing extra libraries

In [3]:
!pip install datasets # This one is for downloading our samsum dataset direclty from Hugging Face
!pip install peft # Both peft and trl are the libs that help us 
!pip install trl # to configure our training methods and params
!pip install bitsandbytes # This one will help us to quantize the model
!pip install mlflow==2.11.0

   ━━━━━━━━━━━━━━━━━━━━━━━━━━━━━━━━━━━━━━━ 510.5/510.5 kB 960.7 kB/s eta 0:00:000:0100:01m
   ━━━━━━━━━━━━━━━━━━━━━━━━━━━━━━━━━━━━━━━ 170.9/170.9 kB 587.6 kB/s eta 0:00:00a 0:00:01
   ━━━━━━━━━━━━━━━━━━━━━━━━━━━━━━━━━━━━━━━━ 1.2/1.2 MB 3.9 MB/s eta 0:00:0000:0100:01m
   ━━━━━━━━━━━━━━━━━━━━━━━━━━━━━━━━━━━━━━━━ 346.4/346.4 kB 3.7 MB/s eta 0:00:00:00:01
   ━━━━━━━━━━━━━━━━━━━━━━━━━━━━━━━━━━━━━━━━ 134.8/134.8 kB 1.2 MB/s eta 0:00:00:00:01
   ━━━━━━━━━━━━━━━━━━━━━━━━━━━━━━━━━━━━━━━ 116.3/116.3 kB 911.7 kB/s eta 0:00:00a 0:00:01
   ━━━━━━━━━━━━━━━━━━━━━━━━━━━━━━━━━━━━━━━━ 194.1/194.1 kB 1.7 MB/s eta 0:00:00ta 0:00:01
   ━━━━━━━━━━━━━━━━━━━━━━━━━━━━━━━━━━━━━━━━ 239.5/239.5 kB 2.9 MB/s eta 0:00:00:00:01
   ━━━━━━━━━━━━━━━━━━━━━━━━━━━━━━━━━━━━━━━━ 124.3/124.3 kB 1.1 MB/s eta 0:00:00:00:01
   ━━━━━━━━━━━━━━━━━━━━━━━━━━━━━━━━━━━━━━━━ 301.6/301.6 kB 2.1 MB/s eta 0:00:00:00:01
  Attempting uninstall: dill
    Found existing installation: dill 0.3.7
    Uninstalling dill-0.3.7:
      Successfully u

### Imports

The libraries used below are already installed by default inside our Deep Learning workspace (excpet for transformers, which was installed together with the extra libraries).

In [4]:
from datasets import load_dataset, Dataset, DatasetDict
import os
import json
import re
from pprint import pprint
import pandas as pd
import torch
from datasets import Dataset, load_dataset
from huggingface_hub import notebook_login
from peft import LoraConfig, PeftModel, AutoPeftModelForCausalLM
from transformers import (
    BitsAndBytesConfig,
    TrainingArguments,
)
from trl import SFTTrainer
import time
import mlflow

2024-03-06 20:00:40.513805: I tensorflow/core/platform/cpu_feature_guard.cc:182] This TensorFlow binary is optimized to use available CPU instructions in performance-critical operations.
To enable the following instructions: AVX2 AVX512F FMA, in other operations, rebuild TensorFlow with the appropriate compiler flags.
2024-03-06 20:00:41.252685: W tensorflow/compiler/tf2tensorrt/utils/py_utils.cc:38] TF-TRT Warning: Could not find TensorRT


### Defining device and model

In [5]:
# 'cuda:0' means that we want to use our GPU, if available. If not, uses CPU.
DEVICE = "cuda:0" if torch.cuda.is_available() else "cpu"
# Model name defines that we are using llama 2 with 7B parameters
# MODEL_NAME = "meta-llama/Llama-2-7b-hf"

### Hugging Face token

In [7]:
HF_TOKEN = "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"

## Dataset

In [8]:
dataset = load_dataset("samsum")

Generating train split:   0%|          | 0/14732 [00:00<?, ? examples/s]

Generating test split:   0%|          | 0/819 [00:00<?, ? examples/s]

Generating validation split:   0%|          | 0/818 [00:00<?, ? examples/s]

In [9]:
def format_dialogue(dialogue):
    # Replace the '\r\n' with '\n' to match the desired output format.
    formatted_dialogue = re.sub(r'\r\n', '\n', dialogue)
    return formatted_dialogue

In [10]:
dataset

DatasetDict({
    train: Dataset({
        features: ['id', 'dialogue', 'summary'],
        num_rows: 14732
    })
    test: Dataset({
        features: ['id', 'dialogue', 'summary'],
        num_rows: 819
    })
    validation: Dataset({
        features: ['id', 'dialogue', 'summary'],
        num_rows: 818
    })
})

This dataset has 3 different divisions:
- Train: has 14732 datapoints
- Test: has 819 datapoints
- Validation: has 818 datapoints

Let's get a sample from the training set and see how it looks like.

In [11]:
dataset['train'][0]

{'id': '13818513',
 'dialogue': "Amanda: I baked  cookies. Do you want some?\r\nJerry: Sure!\r\nAmanda: I'll bring you tomorrow :-)",
 'summary': 'Amanda baked cookies and will bring Jerry some tomorrow.'}

Well, each datapoint is a dict composed by 3 key-values pair:
- The id of the conversation
- The dialogue of the conversation
- The summary of that dialogue

Here's the dialogue from this first datapoint:

In [12]:
dataset['train'][0]['dialogue']

"Amanda: I baked  cookies. Do you want some?\r\nJerry: Sure!\r\nAmanda: I'll bring you tomorrow :-)"

We can see that we have some '\r' between the user's name. We will have to clean up that Rs by replacing them for a '\n', which means we will have a white space before the user's name, lke this:

Amanda: I baked  cookies. Do you want some?
Jerry: Sure!
Amanda: I'll bring you tomorrow :-)

Let's create a function th formats the dialogues for us.
at

In [13]:
def format_dialogue(dialogue):
    # Replace the '\r\n' with '\n' to match the desired output format.
    formatted_dialogue = re.sub(r'\r\n', '\n', dialogue)
    return formatted_dialogue

Now let's test that funtion on the first train dialogue and see what happens.

In [14]:
print(format_dialogue(dataset['train'][0]['dialogue']))

Amanda: I baked  cookies. Do you want some?
Jerry: Sure!
Amanda: I'll bring you tomorrow :-)


Perfect! It works pretty well!

Now it is time to build the input prompt for llama 2 model.

## Defining input prompt

In [15]:
DEFAULT_SYSTEM_PROMPT = """
Below is a conversation between friends in a chat. Write a summary of their conversation.
""".strip()

In [16]:
def generate_training_prompt(
    conversation: str, summary: str, system_prompt: str = DEFAULT_SYSTEM_PROMPT
) -> str:
    return f"""### Instruction: {system_prompt}

### Input:
{conversation.strip()}

### Response:
{summary}
""".strip()

In [17]:
def generate_text(data_point):
    summary = data_point['summary']
    conversation_text = format_dialogue(data_point['dialogue'])
    return {
        "conversation": conversation_text,
        "summary": summary,
        "text": generate_training_prompt(conversation_text, summary),
    }

In [18]:
def process_dataset(data: Dataset):
    return (
        data.shuffle(seed=42)
        .map(generate_text))

## Processing the dataset

In [19]:
# dataset["train"] = process_dataset(dataset["train"])
dataset["train"] = process_dataset(dataset["train"].select(range(1500)))
dataset["test"] = process_dataset(dataset["test"])

Map:   0%|          | 0/1500 [00:00<?, ? examples/s]

Map:   0%|          | 0/819 [00:00<?, ? examples/s]

## Instantiating the model and its tokenizer

In [20]:
notebook_login()

## Quantization, Tokenizer and Model

In [21]:
from transformers import AutoTokenizer, AutoModelForCausalLM

In [22]:
def create_model_and_tokenizer():
    bnb_config = BitsAndBytesConfig( #BitsAndBytes is the reponsable librarie that help us to configure the desired quantization.
        load_in_4bit=True, # In this case, we are loading the model in a 4bir precision instead of its orginal precision.
        bnb_4bit_quant_type="nf4", # normalized float 4 bit data type
        bnb_4bit_compute_dtype=torch.float16, # This is to use half the memory and fit the model
    )

    model = AutoModelForCausalLM.from_pretrained( 
        "google/gemma-2b", # Here we are defining the llama 2 7B model
        use_safetensors=True, #  for storing and loading tensors
        quantization_config=bnb_config, # with the desired quantization 
        trust_remote_code=True, 
        device_map="auto" # and put each layer of the model depending on the available resources
        
    )

    tokenizer = AutoTokenizer.from_pretrained("google/gemma-2b") # Here we tell we want to use the same tokenizer that the model ises
    tokenizer.pad_token = tokenizer.eos_token # Telling that all the padding tokens should be the same as the 'end of sentence'
    tokenizer.padding_side = "right" # and the side padding is the right side

    return model, tokenizer

Great! Let's instatiate both model and tokenizer:

Note: this may take a few minutes to run :)

In [23]:
model, tokenizer = create_model_and_tokenizer()
model.config.use_cache = False # The cache is only used for generation, not for training

config.json:   0%|          | 0.00/627 [00:00<?, ?B/s]

model.safetensors.index.json:   0%|          | 0.00/13.5k [00:00<?, ?B/s]

model-00001-of-00002.safetensors:   0%|          | 0.00/4.95G [00:00<?, ?B/s]

model-00002-of-00002.safetensors:   0%|          | 0.00/67.1M [00:00<?, ?B/s]

Loading checkpoint shards:   0%|          | 0/2 [00:00<?, ?it/s]

generation_config.json:   0%|          | 0.00/137 [00:00<?, ?B/s]

tokenizer_config.json:   0%|          | 0.00/1.11k [00:00<?, ?B/s]

tokenizer.model:   0%|          | 0.00/4.24M [00:00<?, ?B/s]

tokenizer.json:   0%|          | 0.00/17.5M [00:00<?, ?B/s]

special_tokens_map.json:   0%|          | 0.00/555 [00:00<?, ?B/s]

## LoRA

If you are not familiar with LoRA, here's a brief explanation from Hugging Face:

"To make fine-tuning more efficient, LoRA’s approach is to represent the weight updates with two smaller matrices (called update matrices) through low-rank decomposition. These new matrices can be trained to adapt to the new data while keeping the overall number of changes low. The original weight matrix remains frozen and doesn’t receive any further adjustments. To produce the final results, both the original and the adapted weights are combined."

Source: https://huggingface.co/docs/peft/conceptual_guides/lora

Basically, LoRA makes fine-tuning more efficient by drastically reducing the number of trainable parameters and that's why this approach is so commum when fine tuning LLMs.

In [24]:
lora_r = 16 # Ranking of the matrix
lora_alpha = 64 # Scaling facotr
lora_dropout = 0.1
lora_target_modules = [ # Selecting what layers of the model we want to use
    "q_proj",
    "up_proj",
    "o_proj",
    "k_proj",
    "down_proj",
    "gate_proj",
    "v_proj",
]


peft_config = LoraConfig(
    r=lora_r,
    lora_alpha=lora_alpha,
    lora_dropout=lora_dropout,
    target_modules=lora_target_modules,
    bias="none",
    task_type="CAUSAL_LM",
)

More details about LoRA config in this amzing blog post: https://medium.com/@manyi.yim/more-about-loraconfig-from-peft-581cf54643db

## Training (Let's rock!)

### MLFlow setup 
We will be using MLFlow for tracking our training metrics.

Let's define a nme for our experiment so we can easily access it on the Monitor tab later :)

In [25]:
os.environ['MLFLOW_EXPERIMENT_NAME'] = 'gemma2b-summary_task-quant4bit-2-trial'

### Training setup
Here, we will define what are the values for epochs, optmizers, learning rate, evaluation, etc.

For more details, you can check Trainer docs: https://huggingface.co/transformers/v3.0.2/main_classes/trainer.html

In [26]:
training_arguments = TrainingArguments(
    per_device_train_batch_size=4, # The batch size per GPU/TPU core/CPU for training.
    gradient_accumulation_steps=4, # Number of updates steps to accumulate the gradients for, before performing a backward/update pass.
    optim="paged_adamw_32bit", #Adam Optimizer
    logging_steps=20, # Number of update steps between two logs if logging_strategy="steps".
    save_steps = 20,
    learning_rate=1e-3, # The initial learning rate for AdamW optimizer
    fp16=True, # Whether to use fp16 16-bit (mixed) precision training instead of 32-bit training.
    max_grad_norm=0.3, # Maximum gradient norm
    num_train_epochs=3, # Number of epochs
    evaluation_strategy="steps", # Evaluation is done (and logged) every eval_steps
    eval_steps=0.2,
    warmup_ratio=0.05, # Ratio of total training steps used for a linear warmup from 0 to learning_rate
    save_strategy="steps", # Save is done at the end of each steps
    group_by_length=True, # Whether or not to group together samples of roughly the same length in the training dataset 
    output_dir='./gemma/gemma-4bit-2', # Path to save model checkpoints and bins
    report_to="mlflow", # Integration to report the results and logs to
    save_safetensors=True, # Use safetensors saving and loading for state dicts instead of default torch.load and torch.save
    lr_scheduler_type="cosine", # The scheduler type to use.
    seed=42, # Random seed that will be set at the beginning of training. 
) 

In [27]:
trainer = SFTTrainer(
    model=model,
    train_dataset=dataset["train"], # Placing our dataset on train
    eval_dataset=dataset["test"], # and test part
    peft_config=peft_config, #LoRA
    dataset_text_field="text",
    max_seq_length=4096, # Specifies the maximum number of tokens of the input
    tokenizer=tokenizer, # Model's tokenizer we loaded before
    args=training_arguments,
)

Map:   0%|          | 0/1500 [00:00<?, ? examples/s]

Map:   0%|          | 0/819 [00:00<?, ? examples/s]

#### Star of training step

In [ ]:
start_time = time.time()
trainer.train()
end_time = time.time()
training_duration = end_time - start_time
mlflow.log_metric("training_time", training_duration)

2024/03/06 20:07:48 INFO mlflow.tracking.fluent: Experiment with name 'gemma2b-summary_task-quant4bit-2-trial' does not exist. Creating a new experiment.


In [ ]:
trainer.save_model()

## Inference time 😎

Let's load our model using PeftModel from_pretrained method like so:

In [34]:
from peft import PeftModel

In [37]:
model = PeftModel.from_pretrained(model, './gemma/gemma-4bit-2')

Let's create our 5-rows test dataset so we can easily try the model

In [38]:
def generate_prompt(
    conversation: str, system_prompt: str = DEFAULT_SYSTEM_PROMPT
) -> str:
    return f"""### Instruction: {system_prompt}

### Input:
{conversation.strip()}

### Response:
""".strip()

In [39]:
examples = []

for data_point in dataset["test"].select(range(5)):
    summary = data_point['summary']
    conversation = data_point['dialogue']
    examples.append(
        {
            "summary": summary,
            "conversation": conversation,
            "prompt": generate_prompt(conversation),
        }
    )
test_df = pd.DataFrame(examples)

And also a function to help us to tokenize the inputs and use the model 

In [40]:
test_df

summary  \
0  Both Claire and Linda are making curry for din...   
1  Derek and Alyssa make fun of Fergie's performa...   
2  Ann wants to buy Josh's laptop for $200. Josh ...   
3  Matt and Tony want to go to the concert of Bon...   
4   Anastasia sent her new school photos to Darrell.   

                                        conversation  \
0  Claire: <file_photo>\r\nKim: Looks delicious.....   
1  Alyssa: Have you seen Fergie’s national anthem...   
2  Ann: Hi, is the laptop still available?\r\nJos...   
3  Matt: have you heard that Bon Jovi are coming ...   
4  Anastasia: Our new school photos\r\nAnastasia:...   

                                              prompt  
0  ### Instruction: Below is a conversation betwe...  
1  ### Instruction: Below is a conversation betwe...  
2  ### Instruction: Below is a conversation betwe...  
3  ### Instruction: Below is a conversation betwe...  
4  ### Instruction: Below is a conversation betwe...

In [41]:
def summarize(model, text: str):
    inputs = tokenizer(text, return_tensors="pt").to(DEVICE)
    inputs_length = len(inputs["input_ids"][0])
    with torch.inference_mode():
        # Adjust temperature to a more stable value
        outputs = model.generate(**inputs, max_new_tokens=256, temperature=0.7)
    return tokenizer.decode(outputs[0][inputs_length:], skip_special_tokens=True)

### Inferences

In [42]:
import numpy as np
np.random.seed(28)

In [43]:
example = test_df.iloc[0]
print(example.conversation)

Claire: <file_photo>
Kim: Looks delicious...
Linda: No way... Look what I'm cooking right now:
Linda: <file_photo>
Claire: hahahaha 
Kim: Curry dream team
Claire: Enjoy your dinner :*


In [44]:
print(example.summary)

Both Claire and Linda are making curry for dinner. 


In [45]:
print(example.prompt)

### Instruction: Below is a conversation between friends in a chat. Write a summary of their conversation.

### Input:
Claire: <file_photo>
Kim: Looks delicious...
Linda: No way... Look what I'm cooking right now:
Linda: <file_photo>
Claire: hahahaha 
Kim: Curry dream team
Claire: Enjoy your dinner :*

### Response:


In [46]:
%%time
summary = summarize(model, example.prompt)

/opt/conda/lib/python3.10/site-packages/transformers/generation/configuration_utils.py:410: UserWarning: `do_sample` is set to `False`. However, `temperature` is set to `0.7` -- this flag is only used in sample-based generation modes. You should set `do_sample=True` or unset `temperature`.
  warnings.warn(


CPU times: user 32.1 s, sys: 3.12 s, total: 35.2 s
Wall time: 35.3 s


In [47]:
print(summary)


Claire sends Kim and Linda a photo of her cooking. Linda sends Claire a photo of her cooking. Kim enjoys Claire's dinner. Claire and Kim are cooking together. Claire and Kim are curry dream team. Linda enjoys Linda's dinner. Linda is cooking. Linda sends Claire a photo of her cooking. Claire and Linda are cooking together. Linda and Claire are curry dream team. Kim enjoys Claire's dinner. Kim is cooking. Kim sends Claire a photo of her cooking. Claire and Kim are cooking together. Claire and Kim are curry dream team. Linda enjoys Linda's dinner. Linda is cooking. Linda sends Claire a photo of her cooking. Claire and Linda are cooking together. Claire and Linda are curry dream team. Kim enjoys Claire's dinner. Kim is cooking. Kim sends Claire a photo of her cooking. Claire and Kim are cooking together. Claire and Kim are curry dream team. Linda enjoys Linda's dinner. Linda is cooking. Linda sends Claire a photo of her cooking. Claire and Linda are cooking together. Claire and Linda are

You can choose other indexes as well and also try with other seeds!

The major part of the outputs are not going to be very good, but sometimes there are some good outputs

### Inference using HuggignFace

The model is also available on the HuggingFace library :)

In [48]:
from peft import PeftModel, PeftConfig
from transformers import AutoModelForCausalLM

config = PeftConfig.from_pretrained("morgana-rodrigues/gemma-2b-4bit")
model = AutoModelForCausalLM.from_pretrained("google/gemma-2b")
model = PeftModel.from_pretrained(model, "morgana-rodrigues/gemma-2b-4bit")

adapter_config.json:   0%|          | 0.00/688 [00:00<?, ?B/s]

Loading checkpoint shards:   0%|          | 0/2 [00:00<?, ?it/s]

adapter_model.safetensors:   0%|          | 0.00/78.5M [00:00<?, ?B/s]

In [51]:
import os

def get_size(start_path):
    total_size = 0
    for dirpath, dirnames, filenames in os.walk(start_path):
        for f in filenames:
            fp = os.path.join(dirpath, f)
            # skip if it is symbolic link
            if not os.path.islink(fp):
                total_size += os.path.getsize(fp)

    return total_size

# Caminhos dos diretórios do modelo
model_4bit_dir = '/home/jovyan/.cache/huggingface/hub/models--morgana-rodrigues--gemma-2b-4bit'

# Obtém o tamanho do diretório do modelo quantizado de 8 bits
size_quant_8bit = get_size(model_8bit_dir)

print(f"Tamanho do modelo quantizado de 8 bits: {size_quant_8bit} bytes")

Tamanho do modelo quantizado de 8 bits: 78480800 bytes


In [52]:
import sys

# Example variable
my_list = model_4bit_dir
# Checking the size of the variable
size_in_bytes = sys.getsizeof(my_list)

print(f"The variable is using {size_in_bytes} bytes of memory.")

The variable is using 125 bytes of memory.
